File: Diameter of a binary tree/diameter of a binary tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class Pair {
        public int first;
        public int second;
        public Pair(int x, int y) {
            first = x;
            second = y;
        }
    }
    
    public int diameterOfBinaryTree(TreeNode root) {
        Pair p = f(root);
        return p.first;
    }
    
    public Pair f(TreeNode node) {
        if (node == null) {
            return new Pair(0, 0);
        }
        else {
            Pair left = f(node.left);
            Pair right = f(node.right);
            
            int height = Math.max(left.second, right.second) + 1;
            int diameter = Math.max(Math.max(left.first, right.first), left.second + right.second);
            return new Pair(diameter, height);
        }
    }
}

File: Diameter of a binary tree/diameter of a binary tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        pair<int, int> p = f(root);
        return p.first - 1;
    }
    
    pair<int, int> f(TreeNode *node) {
        if (node != NULL) {
            pair<int, int> left = f(node->left);
            pair<int, int> right = f(node->right);
            
            int height = max(left.second, right.second) + 1;
            int diameter = max(max(left.first, right.first), left.second + right.second + 1);
            

            return make_pair(diameter, height);
        }
        else {
            return make_pair(0, 0);
        }
    }
};